### ISSUE TITLE
`build`: one way of reducing sudo waits

### ISSUE THREAD

**Issue by greg-minshall** [AladW/aurutils#1012]
i've been migrating, and re-migrating, to a new machine.  (i run arch linux, in case that turns out to be relevant.)  i've seen various discussions here of how to deal with `sudo` (like #907).

one approach i'm playing with now is to do installation of the AUR packages in a many-phased fashion (using a Makefile).  the input to the whole thing is, of course, all the AUR packages i need from the previous machine.  i then use `aur depends` (or the meta file) to produce the transitive closure of `depends` and `makedepends`, as well as whatever `pacman` packages are needed.  after than, i start my phased install; certainly this list isn't optimal, but here are my current (or, almost current) phases.

0. download all the packages (just the `PKGBUILDs`, whatever `aur-fetch` does)
1. [should do] make sure all keys listed in any `validpgpkeys` in the downloaded `PKGBUILDs` are installed (using `gpg` `--locate-keys` and `--recv-keys`)
2. install all `pacman` packages needed by any of our packages (including `depends`, `makedepends`)
3. [messy -- building and `sudo`]  if any packages that are `makedepends` *themselves* have `makedepends`, build and synchronize those latter `makedepends` to `pacman` (`sudo`) and then `pacman -S` (`sudo`) them  (trying to split this into build/install phases seems problematic, given that some of *these* packages may themselves have `makedepends`)
4. process the remaining `makedepends`:
  - build all the remaining `makedepends`
  - [`sudo`] install these packages by synchronizing the repository to `pacman`, and then `pacman -S` them
5. process the rest of the packages
  - build all the remaining packages
  - [`sudo`] install these packages by synchronizing the repository to `pacman`, and then `pacman -S` them

i know this looks like a hack, but it *seems* to work (in the sense of reducing the number of `sudo` prompts, and, if a `sudo` prompt times out, getting back to the prompt is fairly quick [except maybe in step 2?]), at least on my fairly simple system.  BUT, to make it work, i had to add two new flags to `aur-sync`, one of which needed to be added to `aur-build`.

- `--no-sync`: this flag is added to `aur-sync` and, if specified, passed on to `aur build`, which prevents `aur-build` from (oh, i see it's changed in `10`; congrats on, and thanks for, the release) doing any `pacsync` or `pacman` operations for its packages.  [it is weird to call `aur sync --no-sync`, but `aur-sync` provides a lot of functionality to download and build packages.]
- `--nodeps`: `aur-sync` passes this -- *instead* of `--syncdeps` -- to `aur-build`, which in turn passes it to `makepkg`.  this keeps `makepkg` from worrying about any missing dependencies (that it *would* have fulfilled, had the `--syncdeps` flag had been passed to it).

in addition to those two modifications, it feels to me in my ignorance it would be nice if an `aur-synch-with-pacman` "porcelain" command existed, to do what `aur-build` does at the end (currently using `pacsync`, etc.).

thoughts?

cheers.

**Comment by greg-minshall** [comment]
in case anyone wants to experiment with this, here is a patch for `--no-sync` (in `aur-sync`) and `--nodeps` (in `aur-sync` and `aur-build`).  i guess github won't let me attach a `Makefile`, so i'll put that in here, also.

the `Makefile` is just what i developed for my own use; it isn't likely to work "out of the box" for anyone else.  but, maybe the first few variable definitions will be enough.

- `MY_INSTALL_MACHINE_DIR`: probably the default is fine
- `IGNORES_FILE`: a file in the local directory of packages to *not* install.
- `MY_AUR_PACKAGES`: a file with the packages to be installed
- `AURLIB`: where the aurutils library with my mods lives
- `AUR_REPO`: the name of the local repository `aurutils` uses

here is the patch:
````
From 4f82f0544a1f4993aa00607dbc5bc94be3c77d5c Mon Sep 17 00:00:00 2001
From: Greg Minshall <minshall@umich.edu>
Date: Thu, 14 Jul 2022 19:35:49 +0300
Subject: [PATCH 2/2] lib/aur-build lib/aur-sync: add --nodeps and (to
 aur-sync) --no-sync

---
 lib/aur-build |  4 +++-
 lib/aur-sync  | 26 ++++++++++++++++++--------
 2 files changed, 21 insertions(+), 9 deletions(-)

diff --git a/lib/aur-build b/lib/aur-build
index 5c39ebf..4450a29 100755
--- a/lib/aur-build
+++ b/lib/aur-build
@@ -111,7 +111,7 @@ opt_long=('arg-file:' 'chroot' 'database:' 'force' 'root:' 'sign' 'gpg-sign'
           'verify' 'directory:' 'no-sync' 'pacman-conf:' 'remove' 'pkgver'
           'rmdeps' 'no-confirm' 'no-check' 'ignore-arch' 'log' 'new'
           'makepkg-conf:' 'bind:' 'bind-rw:' 'prevent-downgrade' 'temp'
-          'syncdeps' 'clean' 'namcap' 'checkpkg' 'user:' 'makepkg-args:'
+          'syncdeps' 'nodeps' 'clean' 'namcap' 'checkpkg' 'user:' 'makepkg-args:'
           'margs:' 'buildscript:' 'dry-run')
 opt_hidden=('dump-options' 'ignorearch' 'noconfirm' 'nocheck' 'nosync' 'repo:'
             'results:' 'results-append:')
@@ -178,6 +178,8 @@ while true; do
             makepkg_common_args+=(--rmdeps) ;;
         -s|--syncdeps)
             makepkg_common_args+=(--syncdeps) ;;
+        --nodeps)
+            makepkg_common_args+=(--nodeps) ;;
         # makepkg options (build)
         -C|--clean)
             makepkg_args+=(--clean) ;;
diff --git a/lib/aur-sync b/lib/aur-sync
index d36abac..ffb6522 100755
--- a/lib/aur-sync
+++ b/lib/aur-sync
@@ -12,7 +12,10 @@ AUR_SYNC_USE_NINJA=${AUR_SYNC_USE_NINJA:-0}
 PS4='+(${BASH_SOURCE}:${LINENO}): ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
 
 # default arguments
-build_args=(--clean --syncdeps) depends_args=() repo_args=() view_args=() filter_args=() fetch_args=()
+build_args=(--clean)
+# this argument may be turned into `--nodeps` by the `--nodeps` flag
+build_args_syncdeps=--syncdeps
+depends_args=() fetch_args=() repo_args=() view_args=() filter_args=()
 
 # default options
 build=1 chkver_depth=2 download=1 view=1 provides=1 graph=1 keep_going=1
@@ -106,13 +109,13 @@ fi
 
 opt_short='d:D:k:U:AcfLnorRSTuv'
 opt_long=('bind:' 'bind-rw:' 'database:' 'directory:' 'ignore:' 'root:'
-          'makepkg-conf:' 'pacman-conf:' 'chroot' 'continue' 'force'
-          'ignore-arch' 'log' 'no-confirm' 'no-ver' 'no-graph' 'no-ver-argv'
-          'no-view' 'no-provides' 'no-build' 'rmdeps' 'sign' 'temp' 'upgrades'
-          'pkgver' 'rebuild' 'rebuild-tree' 'rebuild-all' 'ignore-file:'
-          'remove' 'provides-from:' 'new' 'prevent-downgrade' 'verify'
-          'makepkg-args:' 'format:' 'no-check' 'keep-going:' 'user:'
-          'rebase' 'reset' 'ff')
+          'makepkg-conf:' 'no-sync' 'nodeps' 'pacman-conf:'
+          'chroot' 'continue' 'force' 'ignore-arch'
+          'log' 'no-confirm' 'no-ver' 'no-graph' 'no-ver-argv' 'no-view'
+          'no-provides' 'no-build' 'rmdeps' 'sign' 'temp' 'upgrades' 'pkgver'
+          'rebuild' 'rebuild-tree' 'rebuild-all' 'ignore-file:' 'remove'
+          'provides-from:' 'new' 'prevent-downgrade' 'verify' 'makepkg-args:'
+          'format:' 'no-check' 'keep-going:' 'user:' 'rebase' 'reset' 'ff')
 opt_hidden=('dump-options' 'allan' 'ignorearch' 'ignorefile:' 'noconfirm'
             'nover' 'nograph' 'nover-argv' 'noview' 'noprovides' 'nobuild'
             'rebuildall' 'rebuildtree' 'rm-deps' 'gpg-sign' 'margs:' 'nocheck'
@@ -190,6 +193,10 @@ while true; do
             shift; build_args+=(--margs "$1") ;;
         --makepkg-conf)
             shift; build_args+=(--makepkg-conf "$1") ;;
+        --no-sync)
+            build_args+=(--no-sync) ;;
+        --nodeps)
+            build_args_syncdeps=--nodeps ;;
         --pacman-conf)
             shift; build_args+=(--pacman-conf "$1")
             repo_args+=(--config "$1")
@@ -237,6 +244,9 @@ while true; do
     shift
 done
 
+# now, do `--syncdeps` or `--nodeps`, as needed
+build_args+=(${build_args_syncdeps})
+
 # shellcheck disable=SC2174
 mkdir -pm 0700 "${TMPDIR:-/tmp}/aurutils-$UID"
 tmp=$(mktemp -d --tmpdir "aurutils-$UID/$argv0.XXXXXXXX")
-- 
2.37.0
````

and, here the `Makefile`
````
# try to phase in the builds, to try to reduce 'sudo timeouts'
#
# 1. all dependencies not managed by aur (so, presumably by pacman)
#
# 2. the dependencies (all) of all aur-based make dependencies of
# anything being installed
#
# 3. all the aur-based make dependencies
#
# 4. then, all the rest.  presumably, this won't ask for any sudo.
# bakalım.

# where things might be on the new machine
MY_INSTALL_MACHINE_DIR = /var/tmp/toit

# packages in that we do not, in fact, wish to install
IGNORES_FILE = ./ignores-file.txt

# the packages i want to install
MY_AUR_PACKAGES = ${MY_INSTALL_MACHINE_DIR}/pkglist_aur.txt

# use my my version of `aur sync` which, oxymoronicly, allows a
# `--no-sync` flag.  and, a new `--nodeps` flag (also in my `aur-build`)
AURLIB = $(shell ls -dt ${HOME}/src/import/arch-linux/aur/aurutils/lib ${MY_INSTALL_MACHINE_DIR}/lib 2> /dev/null | head -1)

# name of the local repository that `aurutils` controls
AUR_REPO=aurutils

# aur-sync cache -- kind of where everything seems to happen
AUR_SYNC_CACHE = ${HOME}/.cache/aurutils/sync

# provided by the AUR team, the AUR database in one simple .json file
AUR_META_FILE = $(shell ls -t {${HOME}/.cache/pacaur-list,${MY_INSTALL_MACHINE_DIR},/var/tmp}/packages-meta-ext-v1.json.gz 2> /dev/null | head -1)
# https://github.com/AladW/aurutils/issues/1011
AUR_QUERY_RPC = https://server.moson.rocks/rpc
# and, communicate this to lower level
export AUR_QUERY_RPC

# directory where things are built
BDIR = ./built

# which AUR packages i want to install
# all the packages that are dependencies of the above
AUR_DEPENDS_TABLE = ${BDIR}/depends.table
# of which, which are from AUR
DEPENDS_AUR = ${BDIR}/depends.aur
# and which are from Pacman
DEPENDS_PAC = ${BDIR}/depends.pac
# of those which are from AUR, which are *make* dependencies?
MAKE_DEPENDS_AUR = ${BDIR}/make_depends.aur
# and, what are *their* dependencies?
MAKE_DEPENDS_DEPENDS_AUR = ${BDIR}/make_depends_depends.aur

# this is for user to add
AUR_SYNC_EXTRA_FLAGS =
AUR_SYNC_FLAGS = --no-view --no-sync --nodeps ${AUR_SYNC_EXTRA_FLAGS}

PACMAN_S_FLAGS = --needed --noconfirm

all: ${AUR_SYNC_CACHE} ${DEPENDS_AUR} ${DEPENDS_PAC} \
		${MAKE_DEPENDS_AUR} ${MAKE_DEPENDS_DEPENDS_AUR} | built

built:
	mkdir ${BDIR}

${AUR_SYNC_CACHE}:
	mkdir -p ${AUR_SYNC_CACHE}

# this is the table we use.
${AUR_DEPENDS_TABLE}: ${AUR_META} ${IGNORES_FILE} | built
	@echo making ${AUR_DEPENDS_TABLE} > /dev/stderr
	@cat ${MY_AUR_PACKAGES} | \
		egrep -v -f <(grep -v '^#' "${IGNORES_FILE}") | \
		xargs -r aur depends -t > ${AUR_DEPENDS_TABLE}

# which make dependencies are from AUR
${DEPENDS_AUR}: ${AUR_DEPENDS_TABLE} | built
	@echo making ${DEPENDS_AUR} > /dev/stderr
	@aq() { \
		unset names; \
		unset sep; \
		while read -r name; do \
			names="$${names}$${sep}$${name}"; \
			sep="|"; \
		done; \
		if [ -n "$${names}" ]; then \
			zcat "${AUR_META_FILE}" | \
                jq -r --arg search "^($${names})\$$" \
                   '[.[] | select(.Name | test($$search)) ] | .[] | .Name'; \
		fi \
	}; \
	awk '$$5~/Depends/{print $$2}' ${AUR_DEPENDS_TABLE} | \
	aq | \
	sort -u > $@

define AURELS
		(grep -w -f <(sed 's/^/^/;s/$$/$$/' ${DEPENDS_AUR}) || sleep 0)
endef

# which make dependencies are from pacman (really: *not* from AUR)
${DEPENDS_PAC}: ${DEPENDS_AUR} ${AUR_DEPENDS_TABLE} | built
	@echo making ${DEPENDS_PAC} > /dev/stderr
	@awk '$$5!~/Self/{print $$2}' ${AUR_DEPENDS_TABLE} | \
		(grep -v -w -f <(sed 's/^/^/;s/$$/$$/' ${DEPENDS_AUR}) || sleep 0) | \
		sort -u > $@

${MAKE_DEPENDS_AUR}: ${DEPENDS.AUR} | built
	@echo making ${MAKE_DEPENDS_AUR} > /dev/stderr
	@awk '$$5~/MakeDepends/{print $$2}' ${AUR_DEPENDS_TABLE} | \
		${AURELS} > $@

${MAKE_DEPENDS_DEPENDS_AUR}: ${AUR_DEPENDS_TABLE} ${MAKE_DEPENDS_AUR} | built
	@echo making ${MAKE_DEPENDS_DEPENDS_AUR} > /dev/stderr
	@cat ${MAKE_DEPENDS_AUR} | \
		aur depends -t - | \
		awk '$$5!~/Self/{print $$2}' | \
		${AURELS} > $@

# phase 0: *download* all the AUR packages.  no `sudo` problems here,
# and (try to) eliminate random network problems
phase0: ${DEPENDS_PAC} ${AUR_SYNC_CACHE}
	@(awk '$$5~/Self/{print $$2}' ${AUR_DEPENDS_TABLE}; \
		cat ${DEPENDS_AUR}) | \
		sort -u | \
		(cd ${AUR_SYNC_CACHE} && aur fetch -)

# phase 1: scan all PKGBUILDs, and make sure any needed `validpgpkeys`
# are installed.  otherwise, a phase below where things are built will
# fail if a key is missing.
phase1:
	@for pb in ~/.cache/aurutils/sync/*/PKGBUILD; do \
		if grep -q validpgpkeys $${pb}; then \
			unset validpgpkeys; \
			. "$${pb}"; \
			if [ -n "$${validpgpkeys[*]}" ]; then \
				for key in "$${validpgpkeys[@]}"; do \
					echo "$${key}"; \
				done; \
			fi; \
		fi; \
	done | \
	sort -u | \
	while read -r key; do \
		if ! gpg --locate-keys "$${key}" | awk 'END{if (NR == 0) exit 1}'; then \
			gpg --recv-keys "$${key}"; \
		fi \
	done;

# phase 2: install whatever pacman packages are needed for whatever
# AUR packages we need.
# msmtp-mta doesn't allow procmail to install.
phase2: ${DEPENDS_PAC}
	@if pacman -T msmtp-mta; then \
		echo removing msmtp-mta && sudo pacman -R msmtp-mta; \
	fi
	@echo installing packages
	@cat ${DEPENDS_PAC} | grep -v msmtp-mta | xargs -r sudo pacman -S ${PACMAN_S_FLAGS}
	@echo removing postfix, installing msmtp-mta
	@sudo pacman -R postfix && sudo pacman -S ${PACMAN_S_FLAGS} msmtp-mta

# XXX for the `sudo pacman -Syu` below -- really, we'd like to synch
# just from our repository (to avoid system upgrads in the middle of
# the instal process).  but, i don't see that aur-build (which might
# normally do this) allows this as an isolated action. there are old
# pacman feature requests, e.g., https://bugs.archlinux.org/task/15561
# https://bugs.archlinux.org/task/16132 to allow this, but i guess
# they haven't achieved any enthusiasm.  aur-build shows the code one
# would use to do this, but i'm lazy

# phase 3: of the aur packages that are make dependencies, what are
# *their* dependencies?
phase3: ${MAKE_DEPENDS_DEPENDS_AUR}
	@cat ${MAKE_DEPENDS_DEPENDS_AUR} | AUR_EXEC_PATH="${AURLIB}" xargs -r aur sync ${AUR_SYNC_FLAGS}
	@if grep -q . ${MAKE_DEPENDS_DEPENDS_AUR}; then sudo pacsync ${AUR_REPO}; fi
	@cat ${MAKE_DEPENDS_DEPENDS_AUR} | xargs -r sudo pacman -S ${PACMAN_S_FLAGS}

# phase 4: build all other AUR make dependencies of our targets
phase4: ${MAKE_DEPENDS_AUR}
	@cat ${MAKE_DEPENDS_AUR} | AUR_EXEC_PATH="${AURLIB}" xargs -r aur sync ${AUR_SYNC_FLAGS}

# phase 5: now, install all other AUR make dependencies of our targets
phase5: ${MAKE_DEPENDS_AUR}
	@if grep -q . ${MAKE_DEPENDS_AUR}; then sudo pacsync ${AUR_REPO}; fi
	@cat ${MAKE_DEPENDS_AUR} | xargs -r sudo pacman -S ${PACMAN_S_FLAGS}

# phase 6: build our targets and all other dependencies
phase6: ${AUR_DEPENDS_TABLE} ${DEPENDS_AUR}
	(awk '$$5~/Self/{print $$2}' ${AUR_DEPENDS_TABLE}; \
		cat ${DEPENDS_AUR}) | \
		sort -u | \
		AUR_EXEC_PATH="${AURLIB}" xargs -r aur sync ${AUR_SYNC_FLAGS}

# phase 7: synchronize and install our targets and all other
# dependencies.  actually, probably shouldn't *need* to use
# DEPENDS_AUR
phase7: ${AUR_DEPENDS_TABLE} ${DEPENDS_AUR}
	sudo pacsync ${AUR_REPO}
	@(awk '$$5~/Self/{print $$2}' ${AUR_DEPENDS_TABLE}; \
		cat ${DEPENDS_AUR}) | \
		sort -u | \
		xargs -r sudo pacman -S ${PACMAN_S_FLAGS}


clean:
	-rm -f ${AUR_DEPENDS_TABLE} ${DEPENDS_AUR} ${DEPENDS_PAC} \
			${MAKE_DEPENDS_AUR} ${MAKE_DEPENDS_DEPENDS_AUR}
````

**Comment by AladW** [comment]
You're definitely on to something. I usually run `makepkg --rmdeps` so I have `O(N)` `sudo` calls in the number of packages. Package conflicts aside, a phased approach could lead to `O(log N)` `sudo` calls. This assumes part 3. is repeated for each layer of the dependency tree, e.g. `makedepends` of `makedepends` of `makedepends` ...

Note that `makepkg` (and by extension, PKGBUILDs) has no strict separation between `depends` and `makedepends`, and `makepkg --syncdeps` will install both. In particular, when installing a `makedepends` package one should assume all `depends` in that package are also required.

Regarding the suggestions:
* `aur-sync --no-sync` - the oxymoronic nature of the option is not a first, see `aur depends --no-depends`. That said, I'd investigate if `aur sync --no-build` suffices.
* `aur-sync --nodeps` - this can be passed to `makepkg` with `aur-sync --margs=--nodeps`, and overwrites `aur-build --syncdeps`. `aur-sync` puts any `--margs` arguments after the default `build_args`.
* `aur-sync-with-pacman` - I support this idea though for different reasons: it allows to easily make e.g. `sudoers` rules for removing the password prompt on the sync step. I've made an initial draft in perl - the tainted mode is useful when the script is run through setuid wrappers that fail to sanitize the environment (e.g. old versions of doas).

Some remarks about the `Makefile`:
* `validpgpkeys` can be retrieved from .SRCINFO, removing the need to source the PKGBUILD at this stage
* `--no-view` is a bad idea unless you are the maintainer of the AUR packages. I see people do this more often, so I wonder if the `vifm` review process is not sufficiently convenient, or it should be complemented by other mechanisms like an abort or notification on maintainer change (#737).

**Comment by AladW** [comment]
Another point: this could be done relatively easily if `aur-depends` would output the depth of each dependency in the dependency tree. For example:
```
foo 9001 foo foo Self 0
foo 9001 foo bar Depends 0
foo 9001 foo baz MakeDepends 0
bar 667 bar bing Depends 1
bing 42 bing zarg MakeDepends 2
```
However `aur-depends` does not collect dependencies level by level, so simply appending the request number to the `--table` output would not work. `aur-graph` could be extended for this - though data structures in `awk` might be too limited for this.

**Comment by greg-minshall** [comment]
thanks for looking at this.

one thing that maybe i didn't clearly explain: the intent is that (other than the messy step 3, building the `makedepends`) is that the steps that download and/or build packages -- *in theory* -- do *not* do any `sudo`.  and, the steps that *do* do a `sudo` do not do any downloading or building.

the idea there is that any step that is going to run a long time (because of downloading/building) can safely be run "unattended".  then, an hour or two later, come back -- hopefully all the builds succeeded.  and, then do the next step, which should (almost?) immediately do a `sudo`, prompting you for your password, do all the syncs, installs, etc., and finish.  then, you move on to the next download/build step.

so, i *think* this means one needs the `--no-sync` flag for `aur-sync`, as one wants `aur-sync` to invoke `aur-build`  to download and build packages, but not sync anything (i.e., i guess, by invoking `makepkg` with `--nodeps`).

that i can pass `--nodeps` to `aur-build` via `--margs` and that it will override `--syncdeps` is good to know.  is it documented, or community knowledge, that `makepkg` will in fact ignore an earlier `--syncdeps` in favor of a later `--nodeps`?

thanks for the pointer at `.SRCINFO`.  i'm a bit of a newbie with AUR stuff.  i'll convert to it.

in terms of `--no-view`, also being a newbie, i couldn't really figure out the purpose.  i guess the idea is that i shouldn't really trust AUR packages, and therefore should make sure a package i'm about to install seems to be doing safe things?  that makes sense, and i will reform.  :)

the dependency depth is probably doable in `awk`, though i haven't looked at `aur-graph` to really see how it is working; data structures in `awk` are pretty complete.  if you'd like me to look at that, let me know (and, the output you'd like to see).

**Comment by AladW** [comment]
> is it documented, or community knowledge, that makepkg will in fact ignore an earlier --syncdeps in favor of a later --nodeps?

No, I don't think it's documented yet. The man page should mention that `--margs` overrides default arguments.

> the dependency depth is probably doable in awk, though i haven't looked at aur-graph to really see how it is working; data structures in awk are pretty complete. if you'd like me to look at that, let me know (and, the output you'd like to see).

Sure, feel free to take a look if you have the time. :+1:

**Comment by greg-minshall** [comment]
here's some awk code.  i was lazy and i didn't look at `aur-graph`, but just took the `--table` output from `aur-depends`.  the code assigns a "level" number for each package: `0` for those that have no reverse make depends (i.e., the inputs to the process); `N > 0` or higher for those that have reverse make depends of levels up to, but not including `N`.

if the file `my-packages.txt` includes the (`AUR`) packages one is interested in, one per line, then
````
cat my-packages.txt | aur depends --table - | awk -f deplevel.awk | sort -n -k2
````

is supposed to give you a sorted list of levels.  (my own list of packages contains no loops, no long lists of dependencies [2 is the highest], etc.)

the awk code (`deplevel.awk`) is
````
#!/usr/bin/env awk -f

# collect the dependencies
$5 == "MakeDepends" {
    packages[$1]++
    packages[$2]++
    if (debug) print "added", $1, "to packages"
    if (debug) print "added", $2, "to packages"
    mdeps[$1,$2] = 1
    if (debug) print $1, "makedepends on", $2
}

# find all packages (or, for level == 0, desired packages) at a given
# level.

# a package at level zero has no reverse dependencies.

# a package at level N has at least one reverse dependency at level
# N-1, and no reverse depdencies at level N or higher

# (the ",            combined, separate, ..." is the awk "idiom" to
# specify local variables.)
function sweep(level,      combined, separate, pkg, rdep, depends, found, INF, saw) {
    if (debug) print "sweep(" level ")"
    INF = level + 1            # everything is relative
    for (combined in mdeps) {   # "Scanning multidimensional arrays"p
        split(combined, separate, SUBSEP)
        pkg = separate[2]       # this package has
        rdep = separate[1]      # this as a "reverse make depends"
        if (pkg in packages) { # only if this package's level isn't already known
            if (debug) {
                print "pkg", pkg, "; rdep", rdep, "(levels", levels[rdep] ")"
            }
            if (levels[rdep] == "") {
                saw[pkg] = INF # we don't know total story for this pkg yet
            } else if ((saw[pkg] == "") || (levels[rdep] > saw[pkg])) {
                saw[pkg] = levels[rdep]
            }
        }
    }
    found = 0
    for (pkg in packages) {
        if (debug) print pkg, "saw", saw[pkg]
        if (saw[pkg] == "") {
            saw[pkg] = -1       # *no* package depends on this one
        }
        if (saw[pkg] + 1 == level) {
            levels[pkg] = saw[pkg] + 1 # we found one
            delete packages[pkg]   # this package is done
            if (debug) {
                print pkg, "has level", levels[pkg], "; length(packages) =", length(packages)
            }
            found++
        }
    }
    if (debug) print "sweep(" level ") returning", found
    return found
}
               

END {
    # now, compute levels 0 ..
    level = 0
    while (sweep(level)) {
        level++
    }

    # error check -- we *hope* that all packages have been computed.
    # i don't see how this wouldn't be the case.  but, ...
    if (length(packages)) {
        print "error: can't compute for package(s):" > "/dev/stderr"
        for (pkg in packages) {
            print pkg
        }
        exit 2
    }

    # generate the output
    for (pkg in levels) {
        print pkg, levels[pkg]
    }

    exit 0
}
````

**Comment by greg-minshall** [comment]
and, i will (hopefully) look at DAG.

**Comment by greg-minshall** [comment]
btw, the above awk script leaves a few things to be desired.  a script with fewer things left to be desired is available.

one still unresolved issue is how to (efficiently) tell if a candidate package is a `virtual package`, like `sh` or `smtp-server`.  `pactree` is fast, but is not a current dependency:
````
pactree imagemagick | grep " provides " | sed -E 's/^[^[:alpha:]]*//' | cut -f3 -d" " | sort -u
````
(and, the grepping/sedding is a bit of a hack.)

**Comment by greg-minshall** [comment]
the maintainer of `gawk-xml` points out that in a `PKGBUILD` file, [`depends`](https://wiki.archlinux.org/title/PKGBUILD#Dependencies) lists any dependencies needed to *build* the package that are *also* needed to run the package (runtime dependencies).  so, those dependencies will not be listed in `makedepends`.
````
  makedepends

  An array of packages that are only required to build the software.
  The minimum dependency version can be specified in the same format
  as in the depends array. The packages in the depends array are
  implicitly required to build the package, they should not be
  duplicated here. 
````

**Comment by AladW** [comment]
Well, that's one interpretation. PKGBUILD(5) only lists the following:

depends (array)
An array of packages this package depends on to run. [...] If the dependency name appears to be a library (ends with .so), makepkg will try to find a binary that depends on the library in the built package and append the version needed by the binary. Appending the version yourself disables automatic detection.

makedepends (array)
An array of packages this package depends on to build but are not needed at runtime. Packages in this list follow the same format as depends.

As I was testing this with ros-noetic-desktop, I got side-tracked with the newly introduced rate limiting on AUR git repositories. Anyway, my basic idea to implement this is the following:

* Retrieve all source files and inspect them with aur-sync --no-build
* Run aur-depends + deplevel.awk to get dependency layers (this is less than ideal because aur-sync already ran aur-depends with the same arguments)
* For each make/depends layer, build AUR packages, then install them (sudo pacman -S --asdeps) together with repo make/depends assuming there are no package conflicts. To allow for easy removal later, keep a package of which packages have been installed (and possibly removed). Then proceed to the next layer.
* (optional) if some packages fail to build, continue building any that are left on the same layer. This is similar to what ninja(1) does if AUR_SYNC_USE_NINJA is enabled.
* After completing layer 0, remove all installed make/depends.

Regarding provides: aur-graph handles provides, but not the deplevel stuff. Perhaps a new aur-graph could combine the provides and dependency tree functionality, but leave .SRCINFO parsing to a different tool. (that is, the input would be a table or package -> dependency edges).

On Sat, 30 Jul 2022 21:54:02 -0700
greg-minshall ***@***.***> wrote:

> the maintainer of `gawk-xml` points out that in a `PKGBUILD` file, [`depends`](https://wiki.archlinux.org/title/PKGBUILD#Dependencies) lists any dependencies needed to *build* the package that are *also* needed to run the package (runtime dependencies).  so, those dependencies will not be listed in `makedepends`.
> ````
>   makedepends
> 
>   An array of packages that are only required to build the software.
>   The minimum dependency version can be specified in the same format
>   as in the depends array. The packages in the depends array are
>   implicitly required to build the package, they should not be
>   duplicated here. 
> ````
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/AladW/aurutils/issues/1012#issuecomment-1200348358
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>


-- 
alad ***@***.***>

**Comment by greg-minshall** [comment]
that sounds good.  one point that maybe you already had in mind:

> For each make/depends layer, build AUR packages, then install them (sudo pacman -S --asdeps) together with repo make/depends assuming there are no package conflicts.

i would have two separate steps for this (on each layer).  the first -- building the AUR packages -- can take an arbitrary amount of time (downloading, building), but doesn't require `sudo`.  the second -- installing them -- should run very quickly, but does require `sudo` (and, therefore, the user's attention).  thus, the user can walk away during the build, come back and quickly do the install part.  then, walk away again while the next layer builds.  maybe the UI could (optionally) be something like, "wait for a user confirmation after building, but right before installing, each layer".

(thanks for `--asdeps`.)

**Comment by Earnestly** [comment]
The problem with not installing them is if they are dependencies for subsequent builds in the chain.

**Comment by greg-minshall** [comment]
@Earnestly yes.  the procedure i was thinking of is, for each "link" in the chain, do the non-sudo parts first (downloading, patching, building, installing in the local `aurutils` repository).  then, pausing
````
press ENTER to sudo:
````
after which, doing the `sudo pacman -S ...` (or whatever it would be).  then, repeating for the next "link" in the chain.

that way, by the time a package *needs* a (make) dependency, that dependency will have already been installed in a previous "link" of the chain.

**Comment by Earnestly** [comment]
This is something I once wanted to look into: _Directed Acyclic Graphs with Parallel Regions_.

For example to build `vi` one gets this dependency graph:

![vi](https://user-images.githubusercontent.com/521538/182587495-4d52fe57-1ce6-4f11-bea4-0dc1bedeeb2a.png)

Here `filesystem`, `linux-api-system` and `tzdata` can be built in parallel, as can `glibc` and `glibc2` afterwards (not sure what `glibc2` is here).

I'm sure there's probably literature on this form but I didn't dig too deeply.

**Comment by greg-minshall** [comment]
@Earnestly indeed, i think, while not the *optimal* parallelization, one can run all the builds in a single `make/dependency layer` in parallel.  cheers.

**Comment by greg-minshall** [comment]
if you look at [this nice video](https://www.youtube.com/watch?v=GNyQxXw_oMQ) with Brian Kernighan about `awk`, you'll (somewhere) hear him point out that `awk` programs start small, but often get longer than reasonable.

[here](https://bin.disroot.org/?a01aa2eb1f6dc2e4#4y4TTQvyUFmBLURnzq6Ye4XUkEvT6zAjht1xpXvdMXiw) is an example (which will expire in a year).  this purports to produce the DAG.

````
echo ros-noetic-desktop | \
    aur depends --table - | \
    ./depdag.awk SEEDS=ros-noetic-desktop | \
    sort -u | \
    less
````
(in this case, of `ros-noetic-desktop`, the `sort -u` takes a very long time, reducing the output from about 5 million lines to less than 500 thousand lines; patches welcome.)

basically, you feed `depdag.awk` a table from `aur depends` for a given set of AUR packages you wish to install; you also specify those packages as the `SEEDS` variable to `depdag.awk` itself.  if i have a list of AUR packages to install, i might invoke `depdag.awk` like this:

````
cat my-packages.txt | \
    grep -v python-mistletoe | \
    aur depends --table - | \
    ./depdag.awk debug=0 SEEDS="`cat my-packages.txt`" | \
    sort -u | \
    less
````

here's something one might apply to the output file to rank dependencies (and the `seeds`); higher numbers should be built and installed before lower numbers.

````
  echo ros-noetic-desktop | \
      aur depends --table - | \
      ./depdag.awk SEEDS=ros-noetic-desktop | \
      sort -u | \
      awk '
  {
   for (i = 1; i <= NF; i++) {
     level = i;
     pkg = $level;
     if (level > levels[pkg]) {
       levels[pkg] = level
     }
   }
  }

  END {
    for (i in levels) {
      print levels[i], i
    }
  }
  ' | \
      sort -nr | \
      less
````

feedback welcome.  cheers.

**Comment by greg-minshall** [comment]
there's an "optimization" one can do.  if `A` `makedepends` on `B`, and `B` `depends` on `C`, the above will normally output something like
````
% echo -e 'A B x x MakeDepends\nB C x x Depends' | ./depdag.awk SEEDS=A
A B C
````

but, in fact, `C` doesn't need to be available until `B` *runs*, i.e., when `A` is being built.  there is a `LABEL=yes` parameter (`yes` could be anything other than blank or `0`) which prepends a tag before each *dependency* specifying whether that dependency is a `makedepends` (`mdep`, for brevity) or `depends` (`dep`):
````
% echo -e 'A B x x MakeDepends\nB C x x Depends' | ./depdag.awk SEEDS=A LABEL=yes
A mdep B dep C
````
when determining levels, a `depends` dependency can be built at the same level as its nearest `makedepends` ancestor (which, in fact, means, "as its parent").

(there are `LABELWITH` and `LABELSEP` parameters to customize the output.)

then, with a reasonable-sized `awk`-script
````
% echo -e 'A B x x MakeDepends\nB C x x Depends' | \
   ./depdag.awk SEEDS=A LABEL=yes | \
   awk '{ level = 1; for (i = 2; i < NF; i += 2) { if ($i == "mdep") level++; $i = level }; print}'
A 2 B 2 C
````
where the number preceding each *dependency* shows the level for that dependency (higher levels need to be built before lower levels).

if one of the seeds has a `depends` dependency, that will show up at level 1 on *that* line.
````
% echo -e 'A B x x Depends' | \
    ./depdag.awk SEEDS=A LABEL=yes | \
    awk '{ level = 1; for (i = 2; i < NF; i += 2) { if ($i == "mdep") level++; $i = level }; print}'
A 1 B
% echo -e 'A B x x Depends\nC B x x MakeDepends' | \
    ./depdag.awk SEEDS=A,C LABEL=yes |  \
    awk '{ level = 1; for (i = 2; i < NF; i += 2) { if ($i == "mdep") level++; $i = level }; print}'
A 1 B
C 2 B
````

**Comment by Earnestly** [comment]
Here is an example of how to use the `<details>` HTML element to post large documents without it always taking up a lot of room:

<details>
<summary>depdag.awk</summary>

```
#!/usr/bin/env -S awk -f

# XXX this doesn't know how to deal with "virtual provisions"
# (PKGBUILD `provides`), such as `bash` providing `sh`, "smtp-server".
# so, if `sh` is provided by `bash` (all within pacman), we will still
# output `sh`, as if it is *not* in pacman.  if we rely on pactree

# XXX how to deal with version numbers

# 1. if A `makedepends` on B, and B `depends` on C, then, effectively,
#    A `makedepends` on C.

# 2. and, if A is a seed, and A `depends` on B, then B should be
#    (basically) a seed, also.

# so, for 1
# : A B
# : A C
# and, for 2
# : A
# : B

# to implement, when processing `depends`, ellide "package".

# NB: if A both `depends` and `makedepends` on B, B should be treated
# as `makedepends`.  i.e., B needs ot be installed before A is built.



BEGIN {
    # default options

    # do we "memoize", i.e., short cut calculations we've done before,
    # returning the previous results (speeds up processing, at penalty
    # of suffering some possible code bug[s])
    MEMOIZE = 1

    # should we prefix each dependency on a line with a label to
    # differentiate "makedepends" from "depends"
    LABEL = ""
    LABELWITH = "mdep,dep"
    LABELSEP = " "

    # XXX name "our" aurutils respository is known to pacman
    aurutils = "aurutils"

    # separator between dependencies in one chain
    SEP = " "
}

# this is my way of allowing defaults to be set (in the "BEGIN" clause
# above), and variables to be set on the command line using the
# "foo=bar" notation.
NR == 1 {
    # this is the list of seeds, i.e., the roots of the graphs
    if (debug) print "SEEDS: '" SEEDS "'"

    if (! SEEDS) {
        print "need to specify (comma-separated) seeds: -v SEEDS=pkga,pkgb" > "/dev/stderr"
        exit 1
    }

    split(SEEDS, x, "[ ,\n]")
    for (i in x) {
        if (x[i] in seeds) {
            seeds[x[i]]++
        } else {
            seeds[x[i]] = 1
        }
    }

    for (seed in seeds) {
        ndepends[seed] = 0
    }

    initpaclist()

    # process label parameters
    if (LABEL) {
        # if we are labeling, what labels to use: a comma-separated list
        # of two "labels"
        if (!LABELWITH) {
            LABELWITH = "mdep,dep"
        }
        # now, process LABELWITH
        split(LABELWITH, labelwith, ",");
        # portably determine length of labelwith
        len = 0;
        for (i in labelwith) {
            len++;
        }
        if (len != 2) {
            print "invalid \"LABELWITH\" (should be a string of two elements, separated by a comma)" > "/dev/stderr"
            exit(1)
        }
        # set up assoc array for ease of computing
        labels["makedepends"] = labelwith[1]
        labels["depends"] = labelwith[2]

        # how the label should be separated from the dependency in
        # question
        if (!LABELSEP) {
            LABELSEP = " "
        }
    }

}

# collect the dependencies
$5 == "MakeDepends" || $5 == "Depends" {
    if (! ($2 in paclist)) {
        # get all the ndepends set to 0
        if (! ($1 in ndepends)) {
            ndepends[$1] = 0
        }
        if (!($2 in ndepends)) {
            ndepends[$2] = 0
        }
        # find location of this entry XXX this means we "loose track"
        # of duplicates, e.g., a "makedepends" and a "depends" on the
        # same package from the same package.  i'm not sure if this is
        # a feature or a bug.  i don't think i need to do this; but,
        # it's in the code right now.  and, removing it causes
        # "Killed" because of memory consumption (with
        # ros-noetic-desktop).
        if ($1 SUBSEP $2 in location) {
            where = location[$1,$2]
        } else {
            # if not there, enter this dependency for this package
            where = ndepends[$1]++
            depends[$1,where] = $2
            location[$1,$2] = where
        }
        if (debug) print $1, tolower($5), "on", $2, "(" ndeps($1) ")"
        # now, what "flavor" of dependency is this?  favor the
        # `makedepends flavor over all else (i.e., over blank, or
        # `depends`)
        if (($1 SUBSEP where in flavors) && (flavors[$1,where] == "makedepends")) {
            # do nothing (i.e., don't change it
        } else {
            flavors[$1,where] = tolower($5)
        }
        if (debug) print "[" NR "] flavors[" $1 "," where "]", flavors[$1,where], $0
    }
}

# we need to know which packages pacman knows about, that do not come
# from *our* "aurutils" repository (named by the awk variable
# `aurutils`).  initialize this list.
function initpaclist(       cmd, fregex, save) {
    save = $0      # save $0 (restored at end)
    # https://bbs.archlinux.org/viewtopic.php?id=187277
    cmd = "pacman -Ss | grep -v '^ '" # this generates a list of all pacman knows about
    fregex = "^" aurutils "/"
    while ((cmd | getline) > 0) {
        if ($0 !~ fregex) {
            if (debug > 9) print "accepted", $0
            sub("^.*/", "", $1)
            if (debug > 9) print "converted to", $1
            if ($1 in paclist) {
                paclist[$1]++
            } else {
                paclist[$1] = 1
            }
        } else {
            if (debug > 9) print "rejected", $0
        }
    }
    close(cmd)
    if (debug > 9) print "initpaclist done"
    $0 = save
}

function ndeps(package,                      i, result) {
    if (ndepends[package]) {
        result = "ndepends: " ndepends[package]
        for (i = 0; i < ndepends[package]; i++) {
            result = result " " depends[package,i]
        }
        return result
    } else {
        return ""
    }
}

# compute the dependency graph for PACKAGE considered as a FLAVOR
# ("makedepends" or "depends").  the output is a string of
# (newline-separated) lines; each line lists a different dependency
# chain, starting from PACKAGE.  so, if A depends on B, and B in turn
# depends on B' and B'' (which themselves have no dependencies), then
# the output might be
#
# A B B'
#
# A B B''

# (the ", i, computed, result, line, nl, junk" is the awk "idiom"
# to specify local variables.)
function compute(package, level,                          i, computed, result, nl, sep, junk, dep, label) {
    if (debug) print "[" level "] computing", package, ndeps(package)
    if (package in results) {
        result = results[package]
        if (debug) {
            junk = result
            gsub("\n", "\n\t", junk)
            print "[" level "]", "returning result: " junk
        }
        return result
    }
    if (package in paclist) {
        print "paclist[" package "] !!!" > "/dev/stderr"
        exit 4
    }
    result = ""
    if (ndepends[package]) {
        nl = ""
        sep = SEP
        for (i = 0; i < ndepends[package]; i++) {
            dep = depends[package,i]
            if (LABEL) {
                label = labels[flavors[package,i]] LABELSEP
            } else {
                label = ""
            }
            computed = compute(dep, level + 1);
            # need to put the name of *this* package in front of each line
            gsub("\n", "\n" package sep label, computed)
            result = result nl package sep label computed
            nl = "\n"  # subsequent dependency chains are separated by a newline
            sep = SEP
        }
    } else {
        result = package
    }
    if (debug) {
        junk = result
        gsub("\n", "\n\t", junk)
        print "[" level "]", "returning result: " junk
    }
    if (MEMOIZE) {
        results[package] = result
    }
    return result
}

END {
    if (debug) print "seeding"
    if (debug) fflush("/dev/stdout")
    for (seed in seeds) {
        if (debug) print "seed computing", seed
        result = compute(seed, 1)
        if (debug) print "seed printing", seed
        print result
        if (debug) fflush("/dev/stdout")
    }

    exit 0
}
```
</details>

**Comment by greg-minshall** [comment]
@Earnestly thank you, very much!  for future reference, something like
````
<details>
<summary>depdag.awk</summary>

```
#!/usr/bin/env -S awk -f

# XXX this doesn't know how to deal with "virtual provisions"
# (PKGBUILD `provides`), such as `bash`
...
    exit 0
}
```
</details>
````

**Comment by Earnestly** [comment]
The blank lines (`\n\n`) between ` ``` ` and `<summary>` and `</details>` are necessary as otherwise the "code" portion is interpreted strangely as markdown.

**Comment by AladW** [comment]
This looks like some great work. Thanks for your continued efforts on this.

Regarding `provides`, I think the main issue is how to describe something is an actual `provides`. `aur-graph` knows this because it looks directly at .SRCINFO, but when the input stems from `aur depends --table`, things are less clear: any `Depends` or `MakeDepends` could be a package name, provided by a different package name, or both.

Perhaps the `--table` output could include a fifth column, stating which `provides` are available. If there are none, it would default to `pkgname`.

```
XXX how to deal with version numbers
```
It should suffice to split a `depends` on `<`, `>`, `<=`, `>=`, or `=` and use `vercmp` accordingly. Note that `vercmp` would only be run if there is an operator (most dependencies are unversioned and running `vercmp` once per package should be avoided).

**Comment by greg-minshall** [comment]
thanks for the reply.  so, the first thing my code in my `makefile` does is decide which of the dependencies are `pacman` and then `pacman -S` all of those.  so, for `provides`, one "requirement" is to be able to figure that out fairly easily.

how does `aur-graph` know which packages, e.g., "provide" `shell`?  you mention it looks in `.SRCINFO` files.  but, if none of `sh`, `bash`, etc., are downloaded, where does it get this information?  it seems like you have to have the entire databases (both `pacman` and `aur`) to know this.  are there RPC queries that can be used to ask this?  (sorry for my ignorance/inexperience.)

cheers, Greg

**Comment by AladW** [comment]
You're absolutely right that it doesn't handle repo `provides`. It does, however, deal with any `provides` that are part of AUR packages in the dependency chain - e.g. `aurutils-git` would provide `aurutils`.

For a related issue, see https://github.com/AladW/aurutils/issues/837

**Comment by AladW** [comment]
https://github.com/AladW/aurutils/commit/dbdfe61d94cfed96d3780724fde941b05ff87956 added `aur-depends --raw`. This way you can use `aur-format` to get a table with any information you need, including `provides`. For example:
```
aur depends --raw | aur format --union Name -f '%n:%P\t%b\t%v\t%D\t%M\n' --delim=,
```
would have `name:provides1,provides2,...` in the first column and similarly `depends1,depends2,...`, `makedepends1,makedepends2,...` in the last two columns. This includes any versioned dependencies or provides.
